docs/create-packages/native-files-in-net-packages.md

```diff
@@ -0,0 +1,183 @@
+
+Consider a package `Contoso.Native` that provides .NET APIs to `contoso.dll` on Windows, `libcontoso.dylib` on OSX, and `libcontoso.so` on Linux, using `[DllImport("contoso")]`.
+The package contents should be (excluding files required by the [Open Packaging Conventions](https://en.wikipedia.org/wiki/Open_Packaging_Conventions))
+
+### Example 1
+
+If `Contoso.Native` is built as AnyCPU:
+
+```text
+Contoso.Native.1.0.0.nuspec
+ref/net8.0/Contoso.Native.dll
+runtimes/any/lib/net8.0/Contoso.Native.dll
+runtimes/linux-arm64/native/libcontoso.so
+runtimes/linux-x64/native/libcontoso.so
+runtimes/osx-arm64/native/libcontoso.dylib
+runtimes/osx-x64/native/libcontoso.dylib
+runtimes/win-arm64/native/contoso.dll
+runtimes/win-x64/native/contoso.dll
+```
+
+While the `ref/` and `runtimes/any/lib` copies of `Contoso.Native.dll` could be deduplicated into a single `lib/net8.0/Contoso.Native.dll` file, this is not recommended because NuGet will believe this package is compatible with projects using `packages.config`, and these projects will fail at runtime when the native `contoso.dll` is not found.
+
+### Example 2
+
+If `Contoso.Native` has any differences between architectures.
+For example, but not limited to, using `#if` to compile difference code for different RIDs, or using `<PlatformTarget>` (or anything else that causes `-platform:` to be passed to the compiler) with a value other than `AnyCPU`.
+
+```text
+Contoso.Native.1.0.0.nuspec
+ref/net8.0/Contoso.Native.dll
+runtimes/linux-arm64/lib/net8.0/Contoso.Native.dll
+runtimes/linux-arm64/native/libcontoso.so
+runtimes/linux-x64/lib/net8.0/Contoso.Native.dll
+runtimes/linux-x64/native/libcontoso.so
+runtimes/osx-arm64/lib/net8.0/Contoso.Native.dll
+runtimes/osx-arm64/native/libcontoso.dylib
+runtimes/osx-x64/lib/net8.0/Contoso.Native.dll
+runtimes/osx-x64/native/libcontoso.dylib
+runtimes/win-arm64/lib/net8.0/Contoso.Native.dll
+runtimes/win-arm64/native/contoso.dll
+runtimes/win-x64/lib/net8.0/Contoso.Native.dll
+runtimes/win-x64/native/contoso.dll
+```
+
+### Example 3
+
+If `Contoso.Native` has conditional compilation per operating system, but not per CPU architecture.
+
+```text
+Contoso.Native.1.0.0.nuspec
+ref/net8.0/Contoso.Native.dll
+runtimes/linux/lib/net8.0/Contoso.Native.dll
+runtimes/linux-arm64/native/libcontoso.so
+runtimes/linux-x64/native/libcontoso.so
+runtimes/osx/lib/net8.0/Contoso.Native.dll
+runtimes/osx-arm64/native/libcontoso.dylib
+runtimes/osx-x64/native/libcontoso.dylib
+runtimes/win/lib/net8.0/Contoso.Native.dll
+runtimes/win-arm64/native/contoso.dll
+runtimes/win-x64/native/contoso.dll
+```
+
+## Projects Targeting .NET Framework
+
+.NET has evolved over time, causing project and package incompatibility when certain features are used.
+What was described above assumes that the project using the package targets .NET 5 or higher, or .NET Core (`netcoreapp`), which uses the maximum features to simplify the package contents as much as possible.
```

Comment:
.NET Core and .NET 5 have been out of support for more than a year now. Is it worth mentioning those? Isn't `.NET` expressive enough these days?

Reply:
Let's say the proposal is to change it to "What was described above assumes the project using the package targets .NET, not .NET Framework". I feel like that's less clear to customers who are not very experienced with .NET and don't understand the history of the .NET vs .NET Framework name. I can see an argument to remove the .NETCoreApp part, but I don't want to have to do (bi)annual maintenance of these docs to bump .NET 6 to .NET 8 and so on.

I don't know if I'm being overly defensive since I wrote these docs, but I think saying something along the lines of "Features were added to .NET 5 and higher" is fine, even if .net 5 it's out of support 🤷

Also, unlike .NET's docs where you seem to have at least some docs with .NET 6, .NET 7, .NET 8 tabs at the top of the docs page, NuGet doesn't have per-version docs. So we need to document things that make sense regardless of which version of .NET, or Visual Studio, or nuget.exe is being used. So, .NET 8 docs that assume that everything .NET 7 and earlier is irrelevant doesn't work for NuGet, at least in my opinion.